shell/packages/accounts-email-token/token_server.js

```diff
@@ -0,0 +1,163 @@
+  var found = false;
+  user.services.emailToken.tokens.forEach(function (userToken) {
+    if((userToken.algorithm === token.algorithm) &&
+       (userToken.digest === token.digest)) {
+      found = true;
+    }
+  });
+
+  return found;
+};
+var checkToken = Accounts._checkToken;
+
+// Handler to login with a token.
+Accounts.registerLoginHandler("emailToken", function (options) {
+  if (!options.emailToken)
+    return undefined; // don't handle
+
+  options = options.emailToken;
+  check(options, {
+    email: String,
+    token: String
+  });
+
+  var user = Meteor.users.findOne({"services.emailToken.email": options.email});
+
+  if (!user) {
+    console.error("User not found:", options.email);
+    return {
+      error: new Meteor.Error(403, "User not found")
+    };
+  }
+
+  if (!user.services || !user.services.emailToken ||
+      !(user.services.emailToken.tokens)) {
+    console.error("User has no token set:", options.email);
+    return {
+      error: new Meteor.Error(403, "User has no token set")
+    };
+  }
+
+  var token = Accounts._hashToken(options.token);
+  var found = checkToken(
+    user,
+    token
+  );
+
+  if (!found) {
+    console.error("Token not found:", user.services.emailToken);
+    return {
+      error: new Meteor.Error(403, "Token not found")
+    };
+  }
+
+  Meteor.users.update({_id: user._id}, {$pull: {"services.emailToken.tokens": token}});
+  return {
+    userId: user._id
+  };
+});
+
+var Url = Npm.require("url");
+
+var ROOT_URL = Url.parse(process.env.ROOT_URL);
+var HOSTNAME = ROOT_URL.hostname;
+
+var makeTokenUrl = function (email, token) {
+  return process.env.ROOT_URL + "/_emailToken/" + encodeURIComponent(email) + "/" + encodeURIComponent(token);
+};
+
+///
+/// EMAIL VERIFICATION
+///
+var sendTokenEmail = function (email, token) {
+  var options = {
+    to:  email,
+    from: "Sandstorm " + HOSTNAME + " <no-reply@" + HOSTNAME + ">",
+    subject: "Sandstorm token for " + HOSTNAME,
+    text: "A token has requested on your behalf to login to a Sandstorm instance at " + HOSTNAME +
+      ". You will need to either navigate to " + makeTokenUrl(email, token) + " or manually copy " +
+      token + " into the login dropdown."
+  };
+
+  Email.send(options);
+};
+
+///
+/// CREATING USERS
+///
+// returns the user id
+var createAndEmailTokenForUser = function (email) {
+  // Unknown keys allowed, because a onCreateUserHook can take arbitrary
+  // options.
+  check(email, String);
+
+  var user = Meteor.users.findOne({"services.emailToken.email": email});
+  var userId;
+
+  // TODO(someday): make this shorter, and handle requests that try to brute force it.
+  var token = Random.id(12);
+  var tokenObj = Accounts._hashToken(token);
+  tokenObj.createdAt = new Date();
+
+  if (user) {
+    if (user.services.emailToken.tokens && user.services.emailToken.tokens.length > 2) {
+      throw new Meteor.Error(403, "Too many tokens have been made for this user. Please find the " +
```

Comment:
"It looks like we sent a log in email to this address not long ago. Please use the one that was already sent (check your spam folder if you can't find it), or wait a while and try again."